website/members/api/v2/filters.py

```diff
@@ -28,6 +33,62 @@ def get_schema_operation_parameters(self, view):
         ]
 
 
+class FormerMemberFilter(filters.BaseFilterBackend):
+    def filter_queryset(self, request, queryset, view):
+        former = request.query_params.get("former", "false")
+
+        if former == "false":
+            # Filter out former members
+            return (
+                queryset.exclude(membership=None)
+                .filter(
+                    Q(membership__until__isnull=True)
+                    | Q(membership__until__gt=timezone.now().date())
+                )
+                .distinct()
+            )
+        elif former == "true":
+            # Filter out current members
+
+            memberships_query = Q(until__gt=datetime.now()) | Q(until=None)
+            members_query = ~Q(id=None)
+
+            # Filter out all current active memberships
+            memberships_query &= Q(type=Membership.MEMBER) | Q(type=Membership.HONORARY)
+            memberships = Membership.objects.filter(memberships_query)
+            members_query &= ~Q(pk__in=memberships.values("user__pk"))
+
+            memberships_query = Q(type=Membership.MEMBER) | Q(type=Membership.HONORARY)
+            memberships = Membership.objects.filter(memberships_query)
+            all_memberships = Membership.objects.all()
+            # Only keep members that were once members, or are legacy users
```

Comment:
This seems a bit complicated and I don't expect it to scale well: We first query memberships to python, and then filter the members by a _huge_ set of memberships. We can at least simplify it to use an Exists subquery instead of a full query on memberships: `qs.filter(Exists(Membership.objects.filter(<types>, user=OuterRef("pk"), <time>)))`

Reply:
Of course we do the same on the website memberlist, but that's also inefficient if there are many memberships

Reply:
We do not query to python, according to my SQL trace in DJDT. It seems the Django query optimizer is smart enough to turn the queryset into a subquery thanks to laziness.